1. Importing Libraries

In [365]:
# libraries 

import numpy as np, pandas as pd, re, os

In [366]:
# setting display options 
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

2. Reading the data 

In [367]:
PROJECT_DIR = "/Users/abhisheksaurav/Desktop/real_estate_project/"
DATA_DIR = "Data Collection/Data/City/chandigarh/"

In [368]:
def get_data(name):
    file_name = "{}.csv".format(name)
    file_path = os.path.join(PROJECT_DIR, DATA_DIR, file_name)
    return pd.read_csv(file_path)

In [369]:
df = get_data('house')

In [370]:
df.head()

property_name  \
0    5 Bedroom House for sale in Sector 70A Gurgaon   
1  5 Bedroom House for sale in Sector 21A Faridabad   
2  10 Bedroom House for sale in Sushant Lok Phase 1   
3    21 Bedroom House for sale in Sector 54 Gurgaon   
4  12 Bedroom House for sale in Sushant Lok Phase 1   

                                                link           society  \
0  https://www.99acres.com/5-bhk-bedroom-independ...  Bptp Visionnaire   
1  https://www.99acres.com/5-bhk-bedroom-independ...               NaN   
2  https://www.99acres.com/10-bhk-bedroom-indepen...               NaN   
3  https://www.99acres.com/21-bhk-bedroom-indepen...               NaN   
4  https://www.99acres.com/12-bhk-bedroom-indepen...               NaN   

        price              rate                   area  \
0  5.25 Crore   ₹ 20,115/sq.ft.  (242 sq.m.) Plot Area   
1   5.7 Crore  ₹ 105,751/sq.ft.   (50 sq.m.) Plot Area   
2   2.1 Crore   ₹ 38,251/sq.ft.   (51 sq.m.) Plot Area   
3     5 Crore   ₹ 43,066/sq.ft.  (108 sq.m.) Plot Area   
4     3 Crore   ₹ 53,763/sq.ft.   (52 sq.m.) Plot Area   

                  areaWithType      bedRoom      bathroom       balcony  \
0  Plot area 290(242.48 sq.m.)   5 Bedrooms   4 Bathrooms  3+ Balconies   
1   Plot area 539(50.07 sq.m.)   5 Bedrooms   4 Bathrooms   2 Balconies   
2       Plot area 61(51 sq.m.)  10 Bedrooms  10 Bathrooms  3+ Balconies   
3  Plot area 129(107.86 sq.m.)  21 Bedrooms  21 Bathrooms  3+ Balconies   
4    Plot area 62(51.84 sq.m.)  12 Bedrooms  12 Bathrooms  3+ Balconies   

                       additionalRoom  \
0                        Servant Room   
1  Store Room,Pooja Room,Servant Room   
2                        Servant Room   
3                        Servant Room   
4                              Others   

                                       address noOfFloor      facing  \
0    29b, Sector 70A Gurgaon, Gurgaon, Haryana  3 Floors  North-East   
1       Sector 21A Faridabad, Gurgaon, Haryana  2 Floors         NaN   
2        Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   
3          Sector 54 Gurgaon, Gurgaon, Haryana  5 Floors       North   
4  1228, Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   

      agePossession                                    nearbyLocations  \
0   0 to 1 Year Old  ['Good Earth City Center 2', 'Kunskapsskolan I...   
1  5 to 10 Year Old                                                NaN   
2   0 to 1 Year Old  ['Sector 42-43 metro station', 'Huda city cent...   
3   1 to 5 Year Old  ['Sector 53-54 metro station', 'Sector 54 chow...   
4   Within 6 months  ['Sector 42-43 metro station', 'Huda city cent...   

                                         description  \
0  Do you wish to buy an independent house in sec...   
1  Hi, we have an independent house/villa availab...   
2  Monthly rental income is rs1,40,000/- Best opt...   
3  129 sq yd plot size. 5 floors built up , fully...   
4  Best for investment purpose. Fully furnished b...   

                                      furnishDetails  \
0  ['1 Wardrobe', '1 Fan', '1 Exhaust Fan', '1 Ge...   
1  ['1 Water Purifier', '5 Fan', '1 Exhaust Fan',...   
2  ['10 Bed', '3 Fan', '10 Geyser', '2 Light', 'N...   
3  ['1 Water Purifier', '21 Fan', '1 Fridge', '1 ...   
4  ['1 Water Purifier', '1 Fridge', '12 Fan', '1 ...   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
1  ['Private Garden / Terrace', 'Park', 'Visitor ...   
2  ['Maintenance Staff', 'Water Storage', 'Visito...   
3  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
4  ['Maintenance Staff', 'Water Storage', 'Visito...   

                                              rating property_id  
0  ['Environment5 out of 5', 'Lifestyle4 out of 5...   K70037724  
1                                                NaN   E69288322  
2  ['Environment5 out of 5', 'Lifestyle5 out of 5...   F69536898  
3  ['Environment4 out of 5', 'Lifestyle5 out of 5...   R69

3. Preliminary Analysis

3.1 Check Data Types

In [371]:
df.dtypes

property_name      object
link               object
society            object
price              object
rate               object
area               object
areaWithType       object
bedRoom            object
bathroom           object
balcony            object
additionalRoom     object
address            object
noOfFloor          object
facing             object
agePossession      object
nearbyLocations    object
description        object
furnishDetails     object
features           object
rating             object
property_id        object
dtype: object

In [372]:
df.shape

(1095, 21)

In [373]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1095 entries, 0 to 1094
Data columns (total 21 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   property_name    1095 non-null   object
 1   link             1095 non-null   object
 2   society          471 non-null    object
 3   price            968 non-null    object
 4   rate             1054 non-null   object
 5   area             1095 non-null   object
 6   areaWithType     987 non-null    object
 7   bedRoom          987 non-null    object
 8   bathroom         987 non-null    object
 9   balcony          987 non-null    object
 10  additionalRoom   589 non-null    object
 11  address          1031 non-null   object
 12  noOfFloor        967 non-null    object
 13  facing           674 non-null    object
 14  agePossession    987 non-null    object
 15  nearbyLocations  913 non-null    object
 16  description      1036 non-null   object
 17  furnishDetails   743 non-null    

3.2 Checking for Duplicate Values 

In [374]:
(
    df
    .duplicated()
    .sum()
)

np.int64(13)

In [375]:
(
    df 
    .loc[df.duplicated(keep = False)]
    .sort_values(['property_id', 'property_name', 'address'])
)

property_name  \
58       4 Bedroom House for sale in Sector 91 Gurgaon   
830      4 Bedroom House for sale in Sector 91 Gurgaon   
55       3 Bedroom House for sale in Sector 48 Gurgaon   
827      3 Bedroom House for sale in Sector 48 Gurgaon   
53      4 Bedroom House for sale in Sector 112 Gurgaon   
825     4 Bedroom House for sale in Sector 112 Gurgaon   
48      4 Bedroom House for sale in Sector 70A Gurgaon   
820     4 Bedroom House for sale in Sector 70A Gurgaon   
59       4 Bedroom House for sale in Sector 86 Gurgaon   
831      4 Bedroom House for sale in Sector 86 Gurgaon   
56       4 Bedroom House for sale in Sector 48 Gurgaon   
828      4 Bedroom House for sale in Sector 48 Gurgaon   
57       4 Bedroom House for sale in Sector 55 Gurgaon   
829      4 Bedroom House for sale in Sector 55 Gurgaon   
54         6 Bedroom House for sale in Uppals Southend   
826        6 Bedroom House for sale in Uppals Southend   
52          5 Bedroom House for sale in Greenwood City   
824         5 Bedroom House for sale in Greenwood City   
50             3 Bedroom House for sale in Palam Vihar   
822            3 Bedroom House for sale in Palam Vihar   
792           12 Bedroom House for sale in DLF Phase 1   
871           12 Bedroom House for sale in DLF Phase 1   
51             8 Bedroom House for sale in Palam Vihar   
823            8 Bedroom House for sale in Palam Vihar   
49   4 Bedroom House for sale in C Block Sushant Lo...   
821  4 Bedroom House for sale in C Block Sushant Lo...   

                                                  link  \
58   https://www.99acres.com/4-bhk-bedroom-independ...   
830  https://www.99acres.com/4-bhk-bedroom-independ...   
55   https://www.99acres.com/3-bhk-bedroom-independ...   
827  https://www.99acres.com/3-bhk-bedroom-independ...   
53   https://www.99acres.com/4-bhk-bedroom-independ...   
825  https://www.99acres.com/4-bhk-bedroom-independ...   
48   https://www.99acres.com/4-bhk-bedroom-independ...   
820  https://www.99acres.com/4-bhk-bedroom-independ...   
59   https://www.99acres.com/4-bhk-bedroom-independ...   
831  https://www.99acres.com/4-bhk-bedroom-independ...   
56   https://www.99acres.com/4-bhk-bedroom-independ...   
828  https://www.99acres.com/4-bhk-bedroom-independ...   
57   https://www.99acres.com/4-bhk-bedroom-independ...   
829  https://www.99acres.com/4-bhk-bedroom-independ...   
54   https://www.99acres.com/6-bhk-bedroom-independ...   
826  https://www.99acres.com/6-bhk-bedroom-independ...   
52   https://www.99acres.com/5-bhk-bedroom-independ...   
824  https://www.99acres.com/5-bhk-bedroom-independ...   
50   https://www.99acres.com/3-bhk-bedroom-independ...   
822  https://www.99acres.com/3-bhk-bedroom-independ...   
792  https://www.99acres.com/12-bhk-bedroom-indepen...   
871  https://www.99acres.com/12-bhk-bedroom-indepen...   
51   https://www.99acres.com/8-bhk-bedroom-independ...   
823  https://www.99acres.com/8-bhk-bedroom-independ...   
49   https://www.99acres.com/4-bhk-bedroom-independ...   
821  https://www.99acres.com/4-bhk-bedroom-independ...   

                         society       price             rate  \
58     DLF New Town Heights3.7 ★   2.5 Crore   ₹ 7,058/sq.ft.   
830    DLF New Town Heights3.7 ★   2.5 Crore   ₹ 7,058/sq.ft.   
55            Vipul Tatvam Villa  7.25 Crore  ₹ 27,971/sq.ft.   
827           Vipul Tatvam Villa  7.25 Crore  ₹ 27,971/sq.ft.   
53      Experion Windchants3.8 ★  7.99 Crore  ₹ 12,519/sq.ft.   
825     Experion Windchants3.8 ★  7.99 Crore  ₹ 12,519/sq.ft.   
48              Bptp Visionnaire   5.4 Crore  ₹ 20,690/sq.ft.   
820             Bptp Visionnaire   5.4 Crore  ₹ 20,690/sq.ft.   
59   DLF New Town Heights 24.0 ★   2.5 Crore   ₹ 7,058/sq.ft.   
831  DLF New Town Heights 24.0 ★   2.5 Crore   ₹ 7,058/sq.ft.   
56            Vipul Tatvam Villa   8.4 Crore  ₹ 21,000/sq.ft.   
828           Vipul Tatvam Villa   8.4 Crore  ₹ 21,000/sq.ft.   
57                           NaN  3.55 Crore  ₹ 24,500/sq.ft. 

Observation: 
1. There are 1095 rows and 21 columns
2. All columns are "object" types which is wrong. 
3. We will follow the same steps as we did on flats.csv file to clean it 
4. There are 13 duplicate rows which we need to drop
5. we will drop `property_id`, `link` as these columns doesn't contribute anything to our analysis. 

4. Detailed Analysis

In [376]:
df.head()

property_name  \
0    5 Bedroom House for sale in Sector 70A Gurgaon   
1  5 Bedroom House for sale in Sector 21A Faridabad   
2  10 Bedroom House for sale in Sushant Lok Phase 1   
3    21 Bedroom House for sale in Sector 54 Gurgaon   
4  12 Bedroom House for sale in Sushant Lok Phase 1   

                                                link           society  \
0  https://www.99acres.com/5-bhk-bedroom-independ...  Bptp Visionnaire   
1  https://www.99acres.com/5-bhk-bedroom-independ...               NaN   
2  https://www.99acres.com/10-bhk-bedroom-indepen...               NaN   
3  https://www.99acres.com/21-bhk-bedroom-indepen...               NaN   
4  https://www.99acres.com/12-bhk-bedroom-indepen...               NaN   

        price              rate                   area  \
0  5.25 Crore   ₹ 20,115/sq.ft.  (242 sq.m.) Plot Area   
1   5.7 Crore  ₹ 105,751/sq.ft.   (50 sq.m.) Plot Area   
2   2.1 Crore   ₹ 38,251/sq.ft.   (51 sq.m.) Plot Area   
3     5 Crore   ₹ 43,066/sq.ft.  (108 sq.m.) Plot Area   
4     3 Crore   ₹ 53,763/sq.ft.   (52 sq.m.) Plot Area   

                  areaWithType      bedRoom      bathroom       balcony  \
0  Plot area 290(242.48 sq.m.)   5 Bedrooms   4 Bathrooms  3+ Balconies   
1   Plot area 539(50.07 sq.m.)   5 Bedrooms   4 Bathrooms   2 Balconies   
2       Plot area 61(51 sq.m.)  10 Bedrooms  10 Bathrooms  3+ Balconies   
3  Plot area 129(107.86 sq.m.)  21 Bedrooms  21 Bathrooms  3+ Balconies   
4    Plot area 62(51.84 sq.m.)  12 Bedrooms  12 Bathrooms  3+ Balconies   

                       additionalRoom  \
0                        Servant Room   
1  Store Room,Pooja Room,Servant Room   
2                        Servant Room   
3                        Servant Room   
4                              Others   

                                       address noOfFloor      facing  \
0    29b, Sector 70A Gurgaon, Gurgaon, Haryana  3 Floors  North-East   
1       Sector 21A Faridabad, Gurgaon, Haryana  2 Floors         NaN   
2        Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   
3          Sector 54 Gurgaon, Gurgaon, Haryana  5 Floors       North   
4  1228, Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   

      agePossession                                    nearbyLocations  \
0   0 to 1 Year Old  ['Good Earth City Center 2', 'Kunskapsskolan I...   
1  5 to 10 Year Old                                                NaN   
2   0 to 1 Year Old  ['Sector 42-43 metro station', 'Huda city cent...   
3   1 to 5 Year Old  ['Sector 53-54 metro station', 'Sector 54 chow...   
4   Within 6 months  ['Sector 42-43 metro station', 'Huda city cent...   

                                         description  \
0  Do you wish to buy an independent house in sec...   
1  Hi, we have an independent house/villa availab...   
2  Monthly rental income is rs1,40,000/- Best opt...   
3  129 sq yd plot size. 5 floors built up , fully...   
4  Best for investment purpose. Fully furnished b...   

                                      furnishDetails  \
0  ['1 Wardrobe', '1 Fan', '1 Exhaust Fan', '1 Ge...   
1  ['1 Water Purifier', '5 Fan', '1 Exhaust Fan',...   
2  ['10 Bed', '3 Fan', '10 Geyser', '2 Light', 'N...   
3  ['1 Water Purifier', '21 Fan', '1 Fridge', '1 ...   
4  ['1 Water Purifier', '1 Fridge', '12 Fan', '1 ...   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
1  ['Private Garden / Terrace', 'Park', 'Visitor ...   
2  ['Maintenance Staff', 'Water Storage', 'Visito...   
3  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
4  ['Maintenance Staff', 'Water Storage', 'Visito...   

                                              rating property_id  
0  ['Environment5 out of 5', 'Lifestyle4 out of 5...   K70037724  
1                                                NaN   E69288322  
2  ['Environment5 out of 5', 'Lifestyle5 out of 5...   F69536898  
3  ['Environment4 out of 5', 'Lifestyle5 out of 5...   R69

4.1 Society

In [377]:
(
    df
    .society
    .value_counts()
)

society
Vipul Tatvam Villa                                   27
International City by SOBHA Phase 2                  26
Emaar MGF Marbella                                   26
International City by Sobha Phase 1                  23
Unitech Uniworld Resorts                             14
DLF City Plots                                       11
DLF City Plots Phase 2                               11
Unitech Espace                                       11
Bptp Visionnaire                                      9
Eldeco Mansionz                                       9
Uppal Southend4.0 ★                                   9
Not Applicable                                        7
Experion Windchants3.8 ★                              6
Ansals Palam Vihar                                    6
Sushant Lok 1 Builder Floors                          6
DLF New Town Heights 24.0 ★                           5
DLF City Phase 1                                      5
Vatika India Next4.1 ★                  

In [378]:
(
    df 
    .society 
    .apply(lambda name: re.sub(r'\d+(\.\d+)?\s?★', '', str(name)).strip())
    .str
    .lower()
    .replace('nan', "independent")

)

0                                        bptp visionnaire
1                                             independent
2                                             independent
3                                             independent
4                                             independent
5                                             independent
6                                      emaar mgf marbella
7                                  luxury dlf city floors
8                                             independent
9                            sushant lok 1 builder floors
10                                            independent
11                                            independent
12                                 dlf city plots phase 3
13                                            independent
14                                            independent
15                                            independent
16                                 dlf city plots phase 2
17            

4.2 Price

In [379]:
df.shape

(1095, 21)

In [380]:
df = df[df['price'] != 'Price on Request']

In [381]:
def treat_price(x):
    if type(x) == float:
        return x
    else:
        if x[1] == 'Lac':
            return round(float(x[0])/100,2)
        else:
            return round(float(x[0]),2)

In [382]:
(
    df 
    .price
    .str 
    .split(' ')
    .apply(treat_price)
)

0        5.25
1        5.70
2        2.10
3        5.00
4        3.00
5        4.50
6       12.00
7       20.00
8       10.85
9        1.95
10       4.45
11       9.00
12       5.00
13      15.00
14       8.25
15       2.00
16      10.75
17       2.10
18       7.35
19       4.45
20       3.40
21      12.00
22       8.40
23       8.50
24       7.60
25       7.00
26       4.75
27       3.50
28       5.00
29      10.56
30       5.46
31      10.00
32       9.00
33      10.00
34       1.95
35       3.30
36      13.00
37       7.20
38       6.00
39       9.00
40       5.80
41       0.55
42       7.49
43      13.50
44      11.95
45      12.39
46       6.25
47        NaN
48       5.40
49       5.75
50       2.55
51       5.60
52      17.00
53       7.99
54       3.75
55       7.25
56       8.40
57       3.55
58       2.50
59       2.50
60        NaN
61        NaN
62        NaN
63        NaN
64        NaN
65        NaN
66        NaN
67        NaN
68        NaN
69        NaN
70        NaN
71    

4.3 Rate

- replacing the rate colname with price_per_sq_ft

In [383]:
(
    df 
    .rate 
    .str 
    .split("/")
    .str 
    .get(0)
    .str
    .replace("₹", " ")
    .str 
    .replace(",", "")
    .str 
    .strip()
    .astype('float')
)

0        20115.0
1       105751.0
2        38251.0
3        43066.0
4        53763.0
5        39062.0
6        37037.0
7        48889.0
8        38150.0
9        10077.0
10       32963.0
11       27778.0
12       34083.0
13       33200.0
14       30556.0
15       14914.0
16       29861.0
17       23333.0
18       41762.0
19       15451.0
20       12938.0
21       38095.0
22       25926.0
23       26235.0
24       38760.0
25       25926.0
26        9374.0
27       13410.0
28       21124.0
29       32593.0
30       24170.0
31       37037.0
32       28490.0
33      122222.0
34       13542.0
35       19097.0
36       48148.0
37       22222.0
38       22222.0
39       33333.0
40       32222.0
41        7290.0
42       27741.0
43       50000.0
44       43534.0
45       45889.0
46       34722.0
47           NaN
48       20690.0
49       29716.0
50       16378.0
51       17284.0
52       37777.0
53       12519.0
54       25562.0
55       27971.0
56       21000.0
57       24500.0
58        7058

4.4 bedrooms

In [384]:
(
    df
    .bedRoom
    .value_counts()
    )

bedRoom
4 Bedrooms     239
5 Bedrooms     195
3 Bedrooms     148
2 Bedrooms      88
6 Bedrooms      82
9 Bedrooms      45
1 Bedroom       38
8 Bedrooms      34
7 Bedrooms      31
12 Bedrooms     29
10 Bedrooms     21
16 Bedrooms     12
13 Bedrooms      4
18 Bedrooms      2
19 Bedrooms      2
11 Bedrooms      2
20 Bedrooms      1
21 Bedrooms      1
36 Bedrooms      1
34 Bedrooms      1
14 Bedrooms      1
Name: count, dtype: int64

In [385]:
(
    df[df 
    .bedRoom
    .isnull()]
    .sample(5)
    
)

property_name  \
355          2 Bedroom House for sale in Rajiv Nagar   
96    4 Bedroom House for sale in Sector 109 Gurgaon   
85     4 Bedroom House for sale in Sector 66 Gurgaon   
1094   2 Bedroom House for sale in Sector 92 Gurgaon   
90    4 Bedroom House for sale in Sector 109 Gurgaon   

                                                   link  \
355   https://www.99acres.com/2-bhk-bedroom-independ...   
96    https://www.99acres.com/4-bhk-bedroom-independ...   
85    https://www.99acres.com/4-bhk-bedroom-independ...   
1094  https://www.99acres.com/2-bhk-bedroom-independ...   
90    https://www.99acres.com/4-bhk-bedroom-independ...   

                                  society price             rate  \
355                                   NaN   NaN  ₹ 66,666/sq.ft.   
96    International City by Sobha Phase 1   NaN  ₹ 11,100/sq.ft.   
85                     Emaar MGF Marbella   NaN  ₹ 15,700/sq.ft.   
1094             Signature Global City 92   NaN  ₹ 13,034/sq.ft.   
90    International City by Sobha Phase 1   NaN  ₹ 11,100/sq.ft.   

                               area areaWithType bedRoom bathroom balcony  \
355         (7 sq.m.) Built-up Area          NaN     NaN      NaN     NaN   
96        (325 sq.m.) Built-up Area          NaN     NaN      NaN     NaN   
85        (521 sq.m.) Built-up Area          NaN     NaN      NaN     NaN   
1094           (89 sq.m.) Plot Area          NaN     NaN      NaN     NaN   
90    (292-358 sq.m.) Built-up Area          NaN     NaN      NaN     NaN   

     additionalRoom                               address noOfFloor facing  \
355             NaN                                   NaN       NaN    NaN   
96              NaN  Sector 109 Gurgaon, Gurgaon, Haryana       NaN    NaN   
85              NaN   Sector 66 Gurgaon, Gurgaon, Haryana       NaN    NaN   
1094            NaN                                   NaN       NaN    NaN   
90              NaN  Sector 109 Gurgaon, Gurgaon, Haryana       NaN    NaN   

     agePossession                                    nearbyLocations  \
355            NaN                                                NaN   
96             NaN  ['Dwarka sector 21 metro station', 'Pacific D2...   
85             NaN  ['Sector 55-56 Rapid Metro Station', 'HUB 66',...   
1094           NaN                                                NaN   
90             NaN  ['Dwarka sector 21 metro station', 'Pacific D2...   

                                            description furnishDetails  \
355                                                 NaN            NaN   
96    With just 6 villas per acre, international cit...            NaN   
85    Emaar mgf marbella in sector-66, gurgaon by em...            NaN   
1094                                                NaN            NaN   
90    With just 6 villas per acre, international cit...            NaN   

     features                                             rating property_id  
355       NaN                                                NaN         NaN  
96        NaN  ['Environment5 out of 5', 'Lifestyle4 out of 5...   T69934814  
85        NaN  ['Environment3 out of 5', 'Lifestyle4 out of 5...   W70088874  
1094      NaN                                                NaN         NaN  
90        NaN  ['Environment5 out of 5', 'Lifestyle4 out of 5...   V70112044

- where bedRoom column have null value most of the other columns have null values as well. so we will drop these null values 

In [386]:
(
    df 
    .bedRoom 
    .str 
    .split(" ")
    .str 
    .get(0)
    
)

0         5
1         5
2        10
3        21
4        12
5        20
6         5
7        16
8         6
9         3
10        4
11        6
12       13
13        5
14        5
15        6
16        9
17        7
18        4
19        6
20        9
21        5
22        4
23        4
24        9
25        5
26        5
27        9
28        9
29        5
30        4
31        5
32       10
33        4
34        2
35        4
36       12
37        3
38        4
39        5
40        5
41        2
42       12
43       16
44       12
45       16
46        5
47        5
48        4
49        4
50        3
51        8
52        5
53        4
54        6
55        3
56        4
57        4
58        4
59        4
60      NaN
61      NaN
62      NaN
63      NaN
64      NaN
65      NaN
66      NaN
67      NaN
68      NaN
69      NaN
70      NaN
71      NaN
72      NaN
73        5
74        3
75      NaN
76      NaN
77        5
78        5
79      NaN
80      NaN
81      NaN
82      NaN
83  

4.5 bathroom

In [387]:
( 
    df 
    .bathroom 
    .isnull() 
    .sum()
)

np.int64(108)

In [388]:
( 
    df[df 
    .bathroom 
    .isnull()]
    .sample(5)
)

property_name  \
319       3 Bedroom House for sale in Sector 66 Gurgaon   
102      3 Bedroom House for sale in Sector 63A Gurgaon   
198       7 Bedroom House for sale in Sector 11 Gurgaon   
110       3 Bedroom House for sale in Sector 73 Gurgaon   
1093  8 Bedroom House for sale in C Block Sushant Lo...   

                                                   link  \
319   https://www.99acres.com/3-bhk-bedroom-independ...   
102   https://www.99acres.com/3-bhk-bedroom-independ...   
198   https://www.99acres.com/7-bhk-bedroom-independ...   
110   https://www.99acres.com/3-bhk-bedroom-independ...   
1093  https://www.99acres.com/8-bhk-bedroom-independ...   

                            society price              rate  \
319   Emaar MGF The Palm Drive4.1 ★   NaN               NaN   
102               Anant Raj Estates   NaN               NaN   
198                             NaN   NaN  ₹ 197,080/sq.ft.   
110                     DLF Alameda   NaN   ₹ 13,333/sq.ft.   
1093                            NaN   NaN   ₹ 31,008/sq.ft.   

                           area areaWithType bedRoom bathroom balcony  \
319   (353 sq.m.) Built-up Area          NaN     NaN      NaN     NaN   
102   (211 sq.m.) Built-up Area          NaN     NaN      NaN     NaN   
198        (13 sq.m.) Plot Area          NaN     NaN      NaN     NaN   
110   (543 sq.m.) Built-up Area          NaN     NaN      NaN     NaN   
1093      (180 sq.m.) Plot Area          NaN     NaN      NaN     NaN   

     additionalRoom                               address noOfFloor facing  \
319             NaN   Sector 66 Gurgaon, Gurgaon, Haryana       NaN    NaN   
102             NaN  Sector 63A Gurgaon, Gurgaon, Haryana       NaN    NaN   
198             NaN                                   NaN       NaN    NaN   
110             NaN   Sector 73 Gurgaon, Gurgaon, Haryana       NaN    NaN   
1093            NaN                                   NaN       NaN    NaN   

     agePossession                                    nearbyLocations  \
319            NaN  ['Sector 55-56 Rapid Metro Station', 'HUB 66',...   
102            NaN  ['Sector 55-56 Metro Station', 'Paras Trinity ...   
198            NaN                                                NaN   
110            NaN  ['Raheja Mall', 'Captain Chandan Lal Marg', 'B...   
1093           NaN                                                NaN   

                                            description furnishDetails  \
319   Emaar mgf the palm drive in sector-66, gurgaon...            NaN   
102   Anant raj estates in sector-63a gurgaon, gurga...            NaN   
198                                                 NaN            NaN   
110   Dlf alameda in sector-73 gurgaon, gurgaon by d...            NaN   
1093                                                NaN            NaN   

     features                                             rating property_id  
319       NaN  ['Environment3 out of 5', 'Lifestyle4 out of 5...   P69542300  
102       NaN  ['Environment4 out of 5', 'Lifestyle4 out of 5...   W70001488  
198       NaN                                                NaN         NaN  
110       NaN                                                NaN   Y70075228  
1093      NaN                                                NaN         NaN

- where bathroom col have null values there other columns have null values as well. so we will drop these null values 

In [389]:
(
    df 
    .bathroom
    .str 
    .split(' ')
    .str 
    .get(0)
)

0         4
1         4
2        10
3        21
4        12
5        20
6         5
7        16
8         6
9         4
10        3
11        7
12       13
13        6
14        6
15        4
16        9
17        7
18        4
19        6
20        9
21        6
22        4
23        4
24        9
25        5
26        5
27        9
28        9
29        5
30        5
31        6
32       10
33        4
34        2
35        4
36       16
37        3
38        4
39        5
40        5
41        1
42       12
43       16
44       12
45       18
46        5
47        5
48        5
49        4
50        3
51        6
52        8
53        6
54        6
55        4
56        5
57        4
58        4
59        4
60      NaN
61      NaN
62      NaN
63      NaN
64      NaN
65      NaN
66      NaN
67      NaN
68      NaN
69      NaN
70      NaN
71      NaN
72      NaN
73        5
74        3
75      NaN
76      NaN
77        5
78        4
79      NaN
80      NaN
81      NaN
82      NaN
83  

4.6 balcony

In [390]:
(
    df[df 
    .balcony
    .isnull()
    ]
    .sample(5)
)

property_name  \
81       5 Bedroom House for sale in Sector 109 Gurgaon   
1093  8 Bedroom House for sale in C Block Sushant Lo...   
331      3 Bedroom House for sale in Sector 109 Gurgaon   
354   2 Bedroom House for sale in Palam Vihar Extension   
105       4 Bedroom House for sale in Sector 48 Gurgaon   

                                                   link  \
81    https://www.99acres.com/5-bhk-bedroom-independ...   
1093  https://www.99acres.com/8-bhk-bedroom-independ...   
331   https://www.99acres.com/3-bhk-bedroom-independ...   
354   https://www.99acres.com/2-bhk-bedroom-independ...   
105   https://www.99acres.com/4-bhk-bedroom-independ...   

                                  society price             rate  \
81    International City by SOBHA Phase 2   NaN  ₹ 13,623/sq.ft.   
1093                                  NaN   NaN  ₹ 31,008/sq.ft.   
331   International City by Sobha Phase 1   NaN  ₹ 10,783/sq.ft.   
354                                   NaN   NaN   ₹ 6,666/sq.ft.   
105                    Vipul Tatvam Villa   NaN  ₹ 18,750/sq.ft.   

                               area areaWithType bedRoom bathroom balcony  \
81    (467-681 sq.m.) Built-up Area          NaN     NaN      NaN     NaN   
1093          (180 sq.m.) Plot Area          NaN     NaN      NaN     NaN   
331       (293 sq.m.) Built-up Area          NaN     NaN      NaN     NaN   
354          (49 sq.m.) Carpet Area          NaN     NaN      NaN     NaN   
105   (372-441 sq.m.) Built-up Area          NaN     NaN      NaN     NaN   

     additionalRoom                               address noOfFloor facing  \
81              NaN  Sector 109 Gurgaon, Gurgaon, Haryana       NaN    NaN   
1093            NaN                                   NaN       NaN    NaN   
331             NaN  Sector 109 Gurgaon, Gurgaon, Haryana       NaN    NaN   
354             NaN                                   NaN       NaN    NaN   
105             NaN   Sector 48 Gurgaon, Gurgaon, Haryana       NaN    NaN   

     agePossession                                    nearbyLocations  \
81             NaN  ['Dwarka Sector 21 Metro Station', 'Pacific D2...   
1093           NaN                                                NaN   
331            NaN  ['Dwarka sector 21 metro station', 'Pacific D2...   
354            NaN                                                NaN   
105            NaN  ['Athena', 'Star Mall', 'The Sixth Element Sch...   

                                            description furnishDetails  \
81    International city by sobha phase 2 in sector-...            NaN   
1093                                                NaN            NaN   
331   With just 6 villas per acre, international cit...            NaN   
354                                                 NaN            NaN   
105   Vipul tatvam villa in sector-48 gurgaon, gurga...            NaN   

     features                                             rating property_id  
81        NaN  ['Environment5 out of 5', 'Lifestyle4 out of 5...   A70075122  
1093      NaN                                                NaN         NaN  
331       NaN  ['Environment5 out of 5', 'Lifestyle4 out of 5...   E69520100  
354       NaN                                                NaN         NaN  
105       NaN  ['Environment4 out of 5', 'Lifestyle4 out of 5...   U70088438

- where balcony is null there are other columns which are null as well. so we will drop these null values 

In [391]:
(
    df 
    .balcony 
    .str 
    .split(" ")
    .str 
    .get(0)
    .replace({"No": "0"})
    .value_counts()
)

balcony
3+    373
2     223
3     161
0     115
1     105
Name: count, dtype: int64

4.7 addtional room 

In [392]:
(
    df 
    .additionalRoom
    .fillna("not available")
    .str 
    .lower()
)

0                                        servant room
1                  store room,pooja room,servant room
2                                        servant room
3                                        servant room
4                                              others
5                                        servant room
6                  servant room,pooja room,store room
7                                        servant room
8                                          store room
9                                        servant room
10                                         store room
11      study room,servant room,pooja room,store room
12                                         pooja room
13      pooja room,study room,servant room,store room
14      study room,servant room,store room,pooja room
15                              pooja room,study room
16                 servant room,pooja room,store room
17                   store room,pooja room,study room
18                          

4.8 noOfFloor

In [393]:
(
    df
    .noOfFloor
    .isnull()
    .sum()
)

np.int64(128)

In [394]:
(
    df 
    .noOfFloor
    .str 
    .split(" ")
    .str 
    .get(0)
)

0         3
1         2
2         5
3         5
4         5
5         4
6         3
7         4
8         3
9         4
10        2
11        3
12        4
13        3
14        3
15        3
16        3
17        3
18        3
19        3
20        3
21        3
22        2
23      NaN
24        4
25        3
26        1
27        3
28        3
29        3
30        2
31        3
32        3
33        1
34        2
35        3
36        4
37        1
38        3
39        3
40        2
41        1
42        4
43        4
44        4
45        4
46        2
47        3
48        3
49        2
50        2
51      NaN
52        3
53        2
54        3
55        2
56        2
57        2
58        3
59        3
60      NaN
61      NaN
62      NaN
63      NaN
64      NaN
65      NaN
66      NaN
67      NaN
68      NaN
69      NaN
70      NaN
71      NaN
72      NaN
73        3
74        3
75      NaN
76      NaN
77        1
78        3
79      NaN
80      NaN
81      NaN
82      NaN
83  

4.9 area 

In [395]:
(
    (df 
    .price
    .str 
    .split(' ')
    .apply(treat_price))*10000000

/

    df 
    .rate 
    .str 
    .split("/")
    .str 
    .get(0)
    .str
    .replace("₹", " ")
    .str 
    .replace(",", "")
    .str 
    .strip()
    .astype('float')
)

0       2.609993e+03
1       5.390020e+02
2       5.490053e+02
3       1.161009e+03
4       5.580046e+02
5       1.152015e+03
6       3.240003e+03
7       4.090900e+03
8       2.844037e+03
9       1.935100e+03
10      1.349998e+03
11      3.239974e+03
12      1.467007e+03
13      4.518072e+03
14      2.699961e+03
15      1.341022e+03
16      3.600013e+03
17      9.000129e+02
18      1.759973e+03
19      2.880072e+03
20      2.627918e+03
21      3.150020e+03
22      3.239991e+03
23      3.239947e+03
24      1.960784e+03
25      2.699992e+03
26      5.067207e+03
27      2.609993e+03
28      2.366976e+03
29      3.239960e+03
30      2.258999e+03
31      2.700003e+03
32      3.159003e+03
33      8.181833e+02
34      1.439965e+03
35      1.728020e+03
36      2.700008e+03
37      3.240032e+03
38      2.700027e+03
39      2.700027e+03
40      1.800012e+03
41      7.544582e+02
42      2.699975e+03
43      2.700000e+03
44      2.744981e+03
45      2.699993e+03
46      1.800012e+03
47           

5. Cleaning Operations

In [396]:
def clean_data(df):
    return (
        df
        .drop_duplicates()
        .drop(columns = ['property_id', 'link'])
        .rename(columns = {"area": "price_per_sq_ft"})
        .dropna(subset = ['bedRoom', 'bathroom', 'balcony', 'noOfFloor'])
        .where(df['price'] != 'Price on Request')
        .assign(
            society = lambda df_: df_['society']
            .apply(lambda name: re.sub(r'\d+(\.\d+)?\s?★', '', str(name)).strip())
            .str.lower()
            .replace('nan', 'independent'),

            price = lambda df_: df_['price']
            .str 
            .split(" ")
            .apply(treat_price),

            price_per_sq_ft = lambda df_: df_['rate']
            .str 
            .split("/")
            .str 
            .get(0)
            .str
            .replace("₹", "")
            .str 
            .replace(",", "")
            .str 
            .strip()
            .astype('float'), 

            bedRoom = lambda df_: df_['bedRoom'] 
            .str 
            .split(" ")
            .str 
            .get(0)
            .astype('int'), 

            bathroom = lambda df_: df_['bathroom']
            .str 
            .split(' ')
            .str 
            .get(0)
            .astype('int'),

            balcony = lambda df_: df_['balcony']
            .str 
            .split(" ")
            .str 
            .get(0)
            .replace({"No": "0"}),

            additionalRoom = lambda df_: df_['additionalRoom']
            .fillna("not available")
            .str 
            .lower(),

            floorNum = lambda df_: df_['noOfFloor']
            .str 
            .split(" ")
            .str 
            .get(0), 

            area = lambda df_: round((df_["price"]*10000000)/( df_["price_per_sq_ft"]))
            
            
        )
        .drop(columns = ['rate', "noOfFloor"])
        

       
        )
    

In [397]:
house_cleaned = clean_data(df)

In [398]:
house_cleaned.shape

(945, 19)

In [399]:

(
    house_cleaned
    .insert(loc = 1, column= "property type", value = "house")
)

In [400]:
house_cleaned.head()

property_name property type  \
0    5 Bedroom House for sale in Sector 70A Gurgaon         house   
1  5 Bedroom House for sale in Sector 21A Faridabad         house   
2  10 Bedroom House for sale in Sushant Lok Phase 1         house   
3    21 Bedroom House for sale in Sector 54 Gurgaon         house   
4  12 Bedroom House for sale in Sushant Lok Phase 1         house   

            society  price  price_per_sq_ft                 areaWithType  \
0  bptp visionnaire   5.25          20115.0  Plot area 290(242.48 sq.m.)   
1       independent   5.70         105751.0   Plot area 539(50.07 sq.m.)   
2       independent   2.10          38251.0       Plot area 61(51 sq.m.)   
3       independent   5.00          43066.0  Plot area 129(107.86 sq.m.)   
4       independent   3.00          53763.0    Plot area 62(51.84 sq.m.)   

   bedRoom  bathroom balcony                      additionalRoom  \
0        5         4      3+                        servant room   
1        5         4       2  store room,pooja room,servant room   
2       10        10      3+                        servant room   
3       21        21      3+                        servant room   
4       12        12      3+                              others   

                                       address      facing     agePossession  \
0    29b, Sector 70A Gurgaon, Gurgaon, Haryana  North-East   0 to 1 Year Old   
1       Sector 21A Faridabad, Gurgaon, Haryana         NaN  5 to 10 Year Old   
2        Sushant Lok Phase 1, Gurgaon, Haryana        West   0 to 1 Year Old   
3          Sector 54 Gurgaon, Gurgaon, Haryana       North   1 to 5 Year Old   
4  1228, Sushant Lok Phase 1, Gurgaon, Haryana        West   Within 6 months   

                                     nearbyLocations  \
0  ['Good Earth City Center 2', 'Kunskapsskolan I...   
1                                                NaN   
2  ['Sector 42-43 metro station', 'Huda city cent...   
3  ['Sector 53-54 metro station', 'Sector 54 chow...   
4  ['Sector 42-43 metro station', 'Huda city cent...   

                                         description  \
0  Do you wish to buy an independent house in sec...   
1  Hi, we have an independent house/villa availab...   
2  Monthly rental income is rs1,40,000/- Best opt...   
3  129 sq yd plot size. 5 floors built up , fully...   
4  Best for investment purpose. Fully furnished b...   

                                      furnishDetails  \
0  ['1 Wardrobe', '1 Fan', '1 Exhaust Fan', '1 Ge...   
1  ['1 Water Purifier', '5 Fan', '1 Exhaust Fan',...   
2  ['10 Bed', '3 Fan', '10 Geyser', '2 Light', 'N...   
3  ['1 Water Purifier', '21 Fan', '1 Fridge', '1 ...   
4  ['1 Water Purifier', '1 Fridge', '12 Fan', '1 ...   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
1  ['Private Garden / Terrace', 'Park', 'Visitor ...   
2  ['Maintenance Staff', 'Water Storage', 'Visito...   
3  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
4  ['Maintenance Staff', 'Water Storage', 'Visito...   

                                              rating floorNum    area  
0  ['Environment5 out of 5', 'Lifestyle4 out of 5...        3  2610.0  
1                                                NaN        2   539.0  
2  ['Environment5 out of 5', 'Lifestyle5 out of 5...        5   549.0  
3  ['Environment4 out of 5', 'Lifestyle5 out of 5...        5  1161.0  
4  ['Environment5 out of 5', 'Lifestyle5 out of 5...        5   558.0

In [401]:
house_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Index: 945 entries, 0 to 1090
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   property_name    945 non-null    object 
 1   property type    945 non-null    object 
 2   society          945 non-null    object 
 3   price            926 non-null    float64
 4   price_per_sq_ft  926 non-null    float64
 5   areaWithType     945 non-null    object 
 6   bedRoom          945 non-null    int64  
 7   bathroom         945 non-null    int64  
 8   balcony          945 non-null    object 
 9   additionalRoom   945 non-null    object 
 10  address          940 non-null    object 
 11  facing           654 non-null    object 
 12  agePossession    945 non-null    object 
 13  nearbyLocations  830 non-null    object 
 14  description      945 non-null    object 
 15  furnishDetails   713 non-null    object 
 16  features         651 non-null    object 
 17  rating           828

In [402]:
dir_path = "/Users/abhisheksaurav/Desktop/real_estate_project/Data Collection/Data/City/chandigarh/cleaned_data"
file_name = "cleaned_house"
house_cleaned.to_csv("{}/{}.csv".format(dir_path, file_name), index= False)